### DISCUSSION TITLE
Urql and Next 13 Dynamic Data Fetching

### DISCUSSION THREAD

**Discussion by SleeplessOne1917** [urql-graphql/urql#2832] [Q&A]
I'm currently creating a client like so:
```typescript
createClient({
      url: `/api/graphql`,
      fetchOptions: {
          cache: 'no-store'
      }
})
```

I'm doing this in the hopes that Nextjs will know to fetch fresh data on each request per the [Next 13 docs](https://beta.nextjs.org/docs/data-fetching/fetching#dynamic-data-fetching). However, when I use a server component that fetches data using that client, requests aren't made to the client each time there is a request that would load the server component. Is there a way to get urql to do this with Next 13?

**Comment by SleeplessOne1917** [comment]
This ended up being an unrelated problem.

**Reply by focux** [reply]
I'm experiencing the same issue than you, how did you solve it?

**Reply by SleeplessOne1917** [reply]
I wasn't really able to solve it unfortunately.

**Reply by focux** [reply]
I see... then how are you getting new data after reloading the page? I'm starting to think that it might be Next.js bug.

**Reply by SleeplessOne1917** [reply]
I had to fetch from the client instead of the server. You're correct about it being a Next.js issue. See https://github.com/vercel/next.js/issues/42291.

**Reply by focux** [reply]
Hey, if you are still experiencing this issue, I found the cause and it's not Next.js fault. By default, the [request policy](https://formidable.com/open-source/urql/docs/basics/document-caching/#request-policies) of URQL is `cache-first`, so you need to change that to `network-only`, at least on server components, so it fetches fresh data on each render.

**Reply by kitten** [reply]
Just to leave a note on this,
if you're sharing the client between the client-side and server-side, as in, you have one place where you instantiate them to pass them to a provider, I'd instead recommend to first try to not reuse the `Client` between different server-side requests with a pattern like the one seen in `next-urql`: https://github.com/urql-graphql/urql/blob/b80fc5572da03f91d08222955e23461ea9d693b2/packages/next-urql/src/init-urql-client.ts#L17

Otherwise, you'll always run the risk of leaking data between different users.
It's really cheap to instantiate and create new clients in urql, so if you don't have to reuse them, try to always recreate them on the server-side.

On a note of the `requestPolicy`, that's obviously a documented escape hatch, but if you change this policy permanently on the client-side as well, you won't get cached requests at all anymore, and still not isolate the data on the server-side necessarily entirely.

We'd always recommend to treat that global option as a last resort and to instead look at it from the perspective of maximising cache value. That's also why we publish a `requestPolicyExchange` (TTL-based refetches) rather than recommending changing the policy to `cache-and-network` by default

**Reply by focux** [reply]
Hey @kitten, thanks for the insights. I'm actually using two clients for that same reason, however, as far as I know, server and client components can't be in the same file, so the `next-urql` method won't work with server components but you can still create two files with two clients with that same idea.

Regarding the `requestPolicyExchange`, I'm still trying to figure out what works and what does not, using server components and URQL together. For instance, Next.js 13 has some [new features](https://beta.nextjs.org/docs/data-fetching/mutating) to refresh a server component data when you do a mutation, so in this case, if I were to use the `requestPolicyExchange` and the Next.js `router.refresh` function, it's my understanding that that would only work if the TTL has already expired.

Also, Next.js 13 implements a strong cache for fetch requests and some revalidation features and honestly, it's getting me more confused about what is the best strategy when using URQL, whether to use Next.js cache capabilities or URQL. Everything is so new, so I understand that you don't have an example yet, but it'd definitely help.
